src/app/GitUI/Editor/Diff/DiffHighlightService.cs

```diff
@@ -272,63 +280,89 @@ private static void MarkDifference(IDocument document, ISegment lineRemoved,
+            }
         }
+        else
+        {
+            int length = lineRemoved.Length - lineStartOffset;
+            if (length > 0)
+            {
+                yield return CreateDimmedMarker(lineRemoved.Offset + lineStartOffset, length, GetRemovedBackColor());
+            }
+        }
+
+        yield break;
+
+        static TextMarker CreateDimmedMarker(int offset, int length, Color color)
+            => CreateTextMarker(offset, length, ColorHelper.DimColor(ColorHelper.DimColor(color)));
+
+        static TextMarker CreateTextMarker(int offset, int length, Color color)
+            => new(offset, length, TextMarkerType.SolidBlock, color, ColorHelper.GetForeColorForBackColor(color));
+
+        static Color GetAddedBackColor() => AppColor.AnsiTerminalGreenBackNormal.GetThemeColor();
+        static Color GetRemovedBackColor() => AppColor.AnsiTerminalRedBackNormal.GetThemeColor();
     }
 
-    private void AddExtraPatchHighlighting(IDocument document)
+    /// <summary>
+    ///  Matches related removed and added lines in a consecutive block of a patch document and marks identical parts dimmed.
+    /// </summary>
+    private void MarkInlineDifferences(IDocument document)
     {
         int line = 0;
-
         bool found = false;
-        int diffContentOffset;
-        List<ISegment> linesRemoved = GetRemovedLines(document, ref line, ref found);
-        List<ISegment> linesAdded = GetAddedLines(document, ref line, ref found);
-        if (linesAdded.Count == 1 && linesRemoved.Count == 1)
-        {
-            ISegment lineA = linesRemoved[0];
-            ISegment lineB = linesAdded[0];
-            if (lineA.Length > 4 && lineB.Length > 4 &&
-                document.GetCharAt(lineA.Offset + 4) == 'a' &&
-                document.GetCharAt(lineB.Offset + 4) == 'b')
-            {
-                diffContentOffset = 5;
-            }
-            else
-            {
-                diffContentOffset = 4;
-            }
 
-            MarkDifference(document, linesRemoved, linesAdded, diffContentOffset);
-        }
+        // Skip the first pair of removed / added lines, which uses to contain the filenames - but without highlighting
+        _ = GetRemovedLines(document, ref line, ref found);
+        _ = GetAddedLines(document, ref line, ref found);
 
-        // overlap when marking
-        diffContentOffset = 1;
+        // Process the next blocks of removed / added lines and mark in-line differences
+        const int diffContentOffset = 1; // in order to skip the prefixes '-' / '+'
         while (line < document.TotalNumberOfLines)
         {
             found = false;
-            linesRemoved = GetRemovedLines(document, ref line, ref found);
-            linesAdded = GetAddedLines(document, ref line, ref found);
+            List<ISegment> linesRemoved = GetRemovedLines(document, ref line, ref found);
```

Comment:
With the current implementation, there is no reason to run MarkInlineDifferences if not found (and not GetAddedLines() if not found in GetRemovedLines(), that is no mayor thing if only-parse once in my follow up is used).

But the algorithm could be expanded, if there are only few lines between two "removed" blocks (maybe just newlines) and similar for added, the blocks could be used for the context.

This could well be done in a follow-up too.

Git detects "if(EXISTS ${download_directory})" as moved.
There is no dimming on the correct line for "file(LOCK ${download_directory}/../cmake.lock RELEASE)" as moved with indentation is in the second block (dimming on file(READ ${download_directory}.download expected_checksum))

```patch
-
-    if(EXISTS ${download_directory})
-      if(CPM_SOURCE_CACHE)
-        file(LOCK ${download_directory}/../cmake.lock RELEASE)
+    if(EXISTS ${download_directory}
+      AND CPM_ARGS_CUSTOM_CACHE_CHECKSUM_COMMAND
+      AND (CPM_CHECK_CACHE_CHECKSUM
+        OR DEFINED CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE))
+      if (CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE)
+        # Explicit checksum provided, ignore value in .downloaded
+        set(expected_checksum ${CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE})
+      else()
+        file(READ ${download_directory}.download expected_checksum)
+        string(STRIP "${expected_checksum}" expected_checksum)
       endif()
 
+    if(EXISTS ${download_directory})
+      # Directory content is considered OK
+      file(LOCK ${download_directory}/../cmake.lock RELEASE)
       cpm_store_fetch_properties(
 ```

Reply:
I have pushed the evaluation of `found`.
Matching non-contiguous blocks / enlarging blocks is orthogonal to this PR, i.e. should be done in a follow-up.

Reply:
Yes, this was just a discussion
I pushed an experiment with allowing three lines of gap in c1ac2643afebe266143494c1b988701a9066b63c. That could probably be at least 5, it is only with "View complete file" that the parse does not hit a @@ Header.